miniapps/elasticity/materials/neohookean.hpp

```diff
@@ -0,0 +1,230 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_ELASTICITY_MAT_NEOHOOKEAN_HPP
+#define MFEM_ELASTICITY_MAT_NEOHOOKEAN_HPP
+
+#include "general/enzyme.hpp"
+#include "gradient_type.hpp"
+#include "linalg/tensor.hpp"
+#include "mfem.hpp"
+
+using mfem::internal::tensor;
+using mfem::internal::make_tensor;
+
+/**
+ * @brief Neo-Hookean material
+ *
+ * Defines a Neo-Hookean material response. It satisfies the material_type
+ * interface for ElasticityOperator::SetMaterial. This material type allows
+ * choosing the method of derivative calculation in `action_of_gradient`.
+ * Choices include methods derived by hand using symbolic calculation and a
+ * variety of automatically computed gradient applications, like
+ * - Enzyme forward mode
+ * - Enzyme reverse mode
+ * - Dual number type forward mode
+ * - Finite difference mode
+ *
+ * @tparam dim
+ * @tparam gradient_type
+ */
+template <int dim = 3, GradientType gradient_type = GradientType::Symbolic>
+struct NeoHookeanMaterial
+{
+   static_assert(dim == 3, "NeoHookean model only defined in 3D");
+
+   /**
+    * @brief Compute the stress response.
+    *
+    * @param[in] dudx derivative of the displacement
+    * @return
+    */
+   template <typename T>
+   MFEM_HOST_DEVICE tensor<T, dim, dim>
+   stress(const tensor<T, dim, dim> &__restrict__ dudx) const
+   {
+      constexpr auto I = mfem::internal::IsotropicIdentity<dim>();
+      T J = det(I + dudx);
+      T p = -2.0 * D1 * J * (J - 1);
+      auto devB = dev(dudx + transpose(dudx) + dot(dudx, transpose(dudx)));
+      auto sigma = -(p / J) * I + 2 * (C1 / pow(J, 5.0 / 3.0)) * devB;
+      return sigma;
+   }
+
+   /**
+    * @brief A method to wrap the stress calculation into a static function.
+    *
+    * This is necessary for Enzyme to access the class pointer (self).
+    *
+    * @param[in] self
+    * @param[in] dudx
+    * @param[in] sigma
+    * @return stress
+    */
+   MFEM_HOST_DEVICE static void
+   stress_wrapper(NeoHookeanMaterial<dim, gradient_type> *self,
+                  tensor<double, dim, dim> &dudx,
+                  tensor<double, dim, dim> &sigma)
+   {
+      sigma = self->stress(dudx);
+   }
+
+   /**
+    * @brief Compute the gradient.
+    *
+    * This method is used in the ElasticityDiagonalPreconditioner type to
+    * compute the gradient matrix entries of the current quadrature point,
+    * instead of the action.
+    *
+    * @param[in] dudx
+    * @return
+    */
+   MFEM_HOST_DEVICE tensor<double, dim, dim, dim, dim>
+   gradient(tensor<double, dim, dim> dudx) const
```

Comment:
I would use a more physically-motivated name for this method as it is user-facing. Maybe `tangent_stiffness`? Also, mechanics people are used to writing stress as a function of strain, so this would be `d_stress / d_strain`. Any thoughts on what is most applicable @bslazarov ?